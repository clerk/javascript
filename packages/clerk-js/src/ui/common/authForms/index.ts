export * from './Body';
export * from './ButtonSet';
export * from './Footer';
export * from './Header';
export * from './LoadingScreen';
export * from './Separator';
