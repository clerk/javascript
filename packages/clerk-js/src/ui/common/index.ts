export * from './authPropHelpers';
export * from './constants';
export * from './forms';
export * from './redirects';
export * from './verification';
export * from './withRedirectToHome';
export * from './SSOCallback';
export * from './EmailLinkVerify';
export * from './EmailLinkStatusCard';
export * from './Wizard';
export * from './PrintableComponent';
export * from './withOrganizationsEnabledGuard';
export * from './QRCode';
