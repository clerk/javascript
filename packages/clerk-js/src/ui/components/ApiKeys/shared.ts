import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
