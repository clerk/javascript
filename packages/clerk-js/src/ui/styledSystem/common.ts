import { InternalTheme } from './types';

const textVariants = (t: InternalTheme) => {
  const base = {
    WebkitFontSmoothing: t.options.$fontSmoothing,
    fontFamily: 'inherit',
  };

  const smallRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$xs,
    lineHeight: t.lineHeights.$shorter,
  } as const;

  const smallMedium = {
    ...smallRegular,
    fontWeight: t.fontWeights.$medium,
    lineHeight: t.lineHeights.$short,
  } as const;

  const smallBold = {
    ...smallMedium,
    fontWeight: t.fontWeights.$bold,
  } as const;

  const extraSmallRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$2xs,
    letterSpacing: t.letterSpacings.$normal,
    lineHeight: t.lineHeights.$none,
  } as const;

  const regularRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$sm,
    lineHeight: t.lineHeights.$shorter,
  } as const;

  const regularMedium = {
    ...regularRegular,
    fontWeight: t.fontWeights.$medium,
  } as const;

  const largeBold = {
    ...base,
    fontWeight: t.fontWeights.$bold,
    fontSize: t.fontSizes.$md,
    lineHeight: t.lineHeights.$taller,
  } as const;

  const largeMedium = {
    ...largeBold,
    fontWeight: t.fontWeights.$medium,
  };

  const xlargeMedium = {
    ...largeBold,
    fontSize: t.fontSizes.$xl,
  } as const;

  const xxlargeMedium = {
    ...xlargeMedium,
    fontSize: t.fontSizes.$2xl,
  } as const;

  const buttonExtraSmallBold = {
    ...extraSmallRegular,
    fontWeight: t.fontWeights.$bold,
    textTransform: 'uppercase',
    fontFamily: t.fonts.$buttons,
  } as const;

  const buttonSmallRegular = {
    ...smallRegular,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  const buttonRegularRegular = {
    ...regularRegular,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  const buttonRegularMedium = {
    ...regularMedium,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  return {
    buttonExtraSmallBold,
    buttonSmallRegular,
    buttonRegularRegular,
    buttonRegularMedium,
    extraSmallRegular,
    smallRegular,
    smallMedium,
    smallBold,
    regularRegular,
    regularMedium,
    largeMedium,
    largeBold,
    xlargeMedium,
    xxlargeMedium,
  } as const;
};

const fontSizeVariants = (t: InternalTheme) => {
  return {
    xss: { fontSize: t.fontSizes.$2xs },
    xs: { fontSize: t.fontSizes.$xs },
    sm: { fontSize: t.fontSizes.$sm },
  } as const;
};

const borderVariants = (t: InternalTheme, props?: any) => {
  return {
    normal: {
      borderRadius: t.radii.$md,
      border: t.borders.$normal,
      ...borderColor(t, props),
    },
  } as const;
};

const borderColor = (t: InternalTheme, props?: any) => {
  return {
    borderColor: props?.hasError ? t.colors.$danger500 : t.colors.$blackAlpha300,
  } as const;
};

const focusRing = (t: InternalTheme) => {
  return {
    '&:focus': {
      '&::-moz-focus-inner': { border: '0' },
      WebkitTapHighlightColor: 'transparent',
      boxShadow: t.shadows.$focusRing.replace('{{color}}', t.colors.$primary200),
      transitionProperty: t.transitionProperty.$common,
      transitionTimingFunction: t.transitionTiming.$common,
      transitionDuration: t.transitionDuration.$focusRing,
    },
  } as const;
};

const focusRingInput = (t: InternalTheme) => {
  return {
    '&:focus': {
      WebkitTapHighlightColor: 'transparent',
      boxShadow: t.shadows.$focusRingInput.replace('{{color}}', t.colors.$primary200),
      transitionProperty: t.transitionProperty.$common,
      transitionTimingFunction: t.transitionTiming.$common,
      transitionDuration: t.transitionDuration.$focusRing,
      borderColor: t.colors.$primary200,
    },
  } as const;
};

const disabled = (t: InternalTheme) => {
  return {
    '&:disabled,&[data-disabled]': {
      cursor: 'not-allowed',
      pointerEvents: 'none',
      opacity: t.opacity.$disabled,
    },
  } as const;
};

const centeredFlex = (display: 'flex' | 'inline-flex' = 'flex') => ({
  display: display,
  justifyContent: 'center',
  alignItems: 'center',
});

const maxHeightScroller = (t: InternalTheme) =>
  ({
    height: '100%',
    overflowY: 'scroll',
    '::-webkit-scrollbar': {
      background: 'transparent',
      width: '8px',
      height: '8px',
    },
    '::-webkit-scrollbar-thumb': {
      background: t.colors.$blackAlpha500,
    },
    '::-webkit-scrollbar-track': {
      background: 'transparent',
    },
  } as const);

export const common = {
  textVariants,
  fontSizeVariants,
  borderVariants,
  focusRing,
  focusRingInput,
  disabled,
  borderColor,
  centeredFlex,
  maxHeightScroller,
};
