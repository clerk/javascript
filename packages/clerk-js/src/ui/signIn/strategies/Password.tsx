import React from 'react';
import { Form } from '@clerk/shared/components/form';
import { FieldState } from 'ui/common';
import { Control } from '@clerk/shared/components/control';
import { Input } from '@clerk/shared/components/input';

export type PasswordProps = {
  handleSubmit: (e: React.FormEvent<HTMLFormElement>) => Promise<void>;
  password: FieldState<string>;
};

export const Password: React.FC<PasswordProps> = ({
  handleSubmit,
  password,
}) => {
  return (
    <Form
      handleSubmit={handleSubmit}
      submitButtonClassName='cl-sign-in-button'
      submitButtonLabel='Sign in'
    >
      <Control
        key='password'
        htmlFor='password'
        label='Password'
        error={password.error}
      >
        <Input
          id='password'
          type='password'
          name='password'
          value={password.value}
          handleChange={el => password.setValue(el.value || '')}
          autoFocus
        />
      </Control>
    </Form>
  );
};
