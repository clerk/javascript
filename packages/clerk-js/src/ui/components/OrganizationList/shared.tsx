import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
