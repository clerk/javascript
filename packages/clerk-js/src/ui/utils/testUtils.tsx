import { AttributeData, Attributes, Clerk, EnvironmentResource, OAuthProviders, SignInData } from '@clerk/types';
import { jest } from '@jest/globals';
import React from 'react';

import { ComponentContext, EnvironmentProvider } from '../contexts';
import { CoreClerkContextWrapper } from '../contexts/CoreClerkContextWrapper';
import { AppearanceProvider } from '../customizables';
import { FlowMetadataProvider } from '../elements';
import { RouteContext } from '../router';
import { InternalThemeProvider } from '../styledSystem';

const initialConfig = {
  frontendApi: '',
  organization: undefined,
  user: null,
  session: null,
  addListener: jest.fn(),
  client: {
    pathRoot: '/client',
    sessions: [],
    signUp: {
      pathRoot: '/client/sign_ups',
      status: null,
      requiredFields: [],
      optionalFields: [],
      missingFields: [],
      unverifiedFields: [],
      verifications: {
        emailAddress: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        phoneNumber: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        web3Wallet: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        externalAccount: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
        },
      },
      username: null,
      firstName: null,
      lastName: null,
      emailAddress: null,
      phoneNumber: null,
      web3wallet: null,
      hasPassword: false,
      unsafeMetadata: {},
      createdSessionId: null,
      createdUserId: null,
      abandonAt: null,
    },
    signIn: {
      pathRoot: '/client/sign_ins',
      status: null,
      supportedIdentifiers: [],
      supportedFirstFactors: [],
      supportedSecondFactors: [],
      firstFactorVerification: {
        pathRoot: '',
        status: null,
        strategy: null,
        nonce: null,
        externalVerificationRedirectURL: null,
        attempts: null,
        expireAt: null,
        error: null,
        verifiedAtClient: null,
      },
      secondFactorVerification: {
        pathRoot: '',
        status: null,
        strategy: null,
        nonce: null,
        externalVerificationRedirectURL: null,
        attempts: null,
        expireAt: null,
        error: null,
        verifiedAtClient: null,
      },
      identifier: null,
      createdSessionId: null,
      userData: {},
    },
    lastActiveSessionId: null,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    id: 'client_2GLpcZTwM08TAm7RCDFrfekjbCO',
  },
} as any; // initialize config

const initialEnvironmentResource = {
  userSettings: {
    id: undefined,
    social: {
      oauth_google: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_google',
      },
    } as OAuthProviders,
    attributes: {
      email_address: {
        enabled: true,
        required: true,
        name: 'email_address',
        verifications: ['email_code'],
        used_for_first_factor: true,
        first_factors: ['email_code'],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: true,
      } as AttributeData,
      phone_number: {
        enabled: true,
        required: false,
        name: 'phone_number',
        verifications: ['phone_code'],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: true,
        second_factors: ['phone_code'],
        verify_at_sign_up: true,
      } as AttributeData,
      username: {
        enabled: false,
        required: false,
        name: 'username',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      first_name: {
        enabled: true,
        required: true,
        name: 'first_name',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      last_name: {
        enabled: true,
        required: true,
        name: 'last_name',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      password: {
        enabled: true,
        required: true,
        name: 'password',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      web3_wallet: {
        enabled: false,
        required: false,
        name: 'web3_wallet',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      authenticator_app: {
        enabled: false,
        required: false,
        name: 'authenticator_app',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      backup_code: {
        enabled: false,
        required: false,
        name: 'backup_code',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
    } as Attributes,
    signIn: {
      second_factor: {
        required: false,
      },
    } as SignInData,
    signUp: {
      custom_action_required: false,
      disable_hibp: false,
      progressive: true,
    },
    socialProviderStrategies: ['oauth_google'],
    authenticatableSocialStrategies: ['oauth_google'],
    web3FirstFactors: [],
    enabledFirstFactorIdentifiers: ['email_address'],
    instanceIsPasswordBased: true,
    hasValidAuthFactor: true,
    pathRoot: '',
  },
  authConfig: {
    pathRoot: '',
    singleSessionMode: false,
  },
  displayConfig: {
    pathRoot: '',
    id: 'display_config_2GLnHgZkgHLIp5F3ciLxw1jdRrX',
    instanceEnvironmentType: 'production',
    applicationName: 'ClerkProd-(local)',
    theme: {
      buttons: {
        font_color: '#ffffff',
        font_family: '"Source Sans Pro", sans-serif',
        font_weight: '600',
      },
      general: {
        color: '#6c47ff',
        padding: '1em',
        box_shadow: '0 2px 8px rgba(0, 0, 0, 0.2)',
        font_color: '#151515',
        font_family: '"Source Sans Pro", sans-serif',
        border_radius: '0.5em',
        background_color: '#ffffff',
        label_font_weight: '600',
      },
      accounts: {
        background_color: '#ffffff',
      },
    },
    preferredSignInStrategy: 'password',
    logoUrl: 'https://images.clerk.services/clerk/logo.svg',
    faviconUrl: null,
    backendHost: '',
    homeUrl: '',
    signInUrl: '',
    signUpUrl: '',
    userProfileUrl: '',
    afterSignInUrl: '',
    afterSignUpUrl: '',
    afterSignOutOneUrl: '',
    afterSignOutAllUrl: '',
    afterSwitchSessionUrl: '',
    branded: true,
    supportEmail: '',
    clerkJSVersion: null,
    experimental__forceOauthFirst: false,
  },
  isSingleSession: () => false,
  isProduction: () => true,
  onWindowLocationHost: () => true,
} as any as EnvironmentResource;

const createClerkFixture = () => {
  const listeners: any[] = [];

  const client = {};
  const session = {};
  const user = {};
  const organization = {};
  const resources = { client, session, user, organization };
  const addListener = (listener: any) => {
    listeners.push(listener);
    listener(resources);
  };
  const updateMock = () => {
    // how do we create new state?
    const newState = { ...resources };
    listeners.forEach(listener => listener(newState));
  };

  return { mockedClerk: { ...resources, addListener } as Clerk, updateMock };
};

type FParam = {
  withUsername: () => void;
  withEmailAddress: () => void;
  withGoogleOAuth: () => void;
  withAllOAuth: () => void;
  withPhoneCode: () => void;
};

type ConfigFn = (f: FParam) => void;

export const createFixture = (configFn: ConfigFn) => {
  const config = initialConfig;
  const f = {
    withUsername: () => {
      // config.user_settings.username = {
      //   enabled: true,
      //   required: true,
      // };
    },
    withGoogleOauth: () => {
      config.user_settings.social.google_oauth = {
        enabled: true,
        required: true,
      };
    },
  } as any as FParam;
  if (configFn) {
    configFn(f);
  }

  const { mockedClerk, updateMock } = createClerkFixture();

  const MockClerkProvider = (props: any) => {
    const { children } = props;
    return (
      <CoreClerkContextWrapper clerk={config}>
        <EnvironmentProvider value={initialEnvironmentResource}>
          <RouteContext.Provider
            value={{
              basePath: '',
              startPath: '',
              fullPath: '',
              indexPath: '',
              currentPath: '',
              queryString: '',
              queryParams: {},
              getMatchData: jest.fn(),
              matches: jest.fn(),
              baseNavigate: jest.fn(),
              navigate: jest.fn(),
              resolve: jest.fn(() => ({ href: 'test.host' } as any)),
              refresh: jest.fn(),
              params: {},
            }}
          >
            <AppearanceProvider appearanceKey={'signIn'}>
              <FlowMetadataProvider flow={'signIn'}>
                <InternalThemeProvider>
                  <ComponentContext.Provider value={{ componentName: 'SignIn' }}>{children}</ComponentContext.Provider>
                </InternalThemeProvider>
              </FlowMetadataProvider>
            </AppearanceProvider>
          </RouteContext.Provider>
        </EnvironmentProvider>
      </CoreClerkContextWrapper>
    );
  };

  return { MockClerkProvider, updateMock };
};
