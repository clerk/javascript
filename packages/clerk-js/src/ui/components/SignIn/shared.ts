import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
