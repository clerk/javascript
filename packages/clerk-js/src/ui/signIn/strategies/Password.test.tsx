import { render, renderJSON, screen, userEvent, waitFor } from '@clerk/shared/testUtils';
import React from 'react';
import { FieldState } from 'ui/common';

import { Password } from './Password';

describe('<Password/>', () => {
  const passwordInput = {
    error: undefined,
    name: 'password',
    value: 'Pa$$word!3',
    setValue: jest.fn(),
    setError: jest.fn(),
  } as FieldState<string>;

  it('renders the password form component', async () => {
    const tree = renderJSON(
      <Password
        handleSubmit={jest.fn()}
        password={passwordInput}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders the password form, enters a password and presses the button', async () => {
    const mockSubmit = jest.fn();

    render(
      <Password
        handleSubmit={mockSubmit}
        password={passwordInput}
      />,
    );

    const inputField = screen.getByLabelText('Password');
    await userEvent.type(inputField, 'SADSD!@D23');

    const button = screen.getByRole('button', { name: /Sign in/i });
    await userEvent.click(button);

    await waitFor(() => {
      expect(mockSubmit).toHaveBeenCalledTimes(1);
      // TODO: Check for the value
      // expect(mockSubmit).toHaveBeenCalledWith({})
    });
  });
});
