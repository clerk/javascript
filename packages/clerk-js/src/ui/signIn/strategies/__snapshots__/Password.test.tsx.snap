// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Password/> renders the password form component 1`] = `
<form
  className="form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <fieldset>
    <div
      className="control"
    >
      <label
        className="label labelContainer"
        htmlFor="password"
      >
        <input
          autoFocus={true}
          className="input"
          id="password"
          name="password"
          onChange={[Function]}
          type="password"
          value="Pa$$word!3"
        />
        <span
          className="label"
        >
          Password
        </span>
      </label>
      <span
        className="error"
      />
    </div>
  </fieldset>
  <fieldset
    className="group"
    disabled={false}
  >
    <button
      className="button solid primary cl-sign-in-button"
      disabled={false}
      style={
        Object {
          "display": "flex",
        }
      }
    >
      Sign in
    </button>
  </fieldset>
</form>
`;
